File: swift/hashing/hashmap.swift
class Pair {
  var key: String
  var val: String

  init(_ key: String, _ val: String) {
    self.key = key
    self.val = val
  }
}

class HashMap {
  var size = 0
  var capacity = 2
  var map: [Pair?] = [nil, nil]
  
  func hash(_ key: String) -> Int {
    var index = 0
    for c in key {
      index += Int(c.asciiValue!)
    }
    return index % capacity
  }

  func get(_ key: String) -> String? {
    var index = hash(key)

    while map[index] != nil {
      if map[index]?.key == key {
        return map[index]?.val
      }
      index += 1
      index = index % capacity
    }
    return nil
  }

  func put(_ key: String, _ val: String) {
    var index = hash(key)

    while true {
      if map[index] == nil {
        map[index] = Pair(key, val)
        size += 1
        if size >= capacity / 2 {
          rehash()
        }
        return
      }
      else if map[index]?.key == key {
        map[index]?.val = val
        return
      }
      index += 1
      index = index % capacity
    }
  }

  func remove(_ key: String) {
    if get(key) == nil {
      return
    }

    var index = hash(key)
    while true {
      if map[index]?.key == key {
        // Removing an element using open-addressing actually causes a bug,
        // because we may create a hole in the list, and our get() may
        // stop searching early when it reaches this hole
        map[index] = nil
        size -= 1
        return
      }
      index += 1
      index = index % capacity
    }
  }

  func rehash() {
    capacity = 2 * capacity
    var newMap: [Pair?] = []
    for _ in 0..<capacity {
      newMap.append(nil)
    }

    let oldMap = map
    map = newMap
    size = 0
    for pair in oldMap {
      if pair != nil {
        put(pair!.key, pair!.val)
      }
    }
  }

  func print() {
    for pair in map {
      if pair != nil {
        Swift.print(pair!.key + " " + pair!.val)
      }
    }
  }
}

File: csharp/hashing/HashMap.cs
using System;
using System.Collections;

/* public class Pair
{
    public string Key { get; set; }
    public string Val { get; set; }

    public Pair(string key, string val)
    {
        this.Key = key;
        this.Val = val;
    }
} */

public class HashMap
{
    int Size { get; set; }
    int Capacity { get; set; }
    Pair[] Map { get; set; }

    public HashMap()
    {
        Size = 0;
        Capacity = 2;
        Map = new Pair[2];
    }

    public int Hash(string key)
    {
        int index = 0;
        for (int i = 0; i < key.Length; i++)
        {
            index += (int)key[i];
        }
        return index % this.Capacity;
    }

    public string Get(string key)
    {
        int index = this.Hash(key);
        while (this.Map[index] != null)
        {
            if (this.Map[index].Key == key)
            {
                return this.Map[index].Val;
            }
            index += 1;
            index %= this.Capacity;
        }
        return string.Empty;
    }

    public void Put(string key, string val)
    {
        int index = this.Hash(key);

        while (true)
        {
            if (this.Map[index] == null)
            {
                this.Map[index] = new Pair(key, val);
                this.Size += 1;
                if (this.Size >= this.Capacity / 2)
                {
                    this.ReHash();
                }
                return;
            }
            else if (this.Map[index].Key == key)
            {
                this.Map[index].Val = val;
                return;
            }
            index += 1;
            index %= this.Capacity;
        }
    }

    public void Remove(string key)
    {
        if (this.Get(key) == null)
        {
            return;
        }

        int index = this.Hash(key);
        while (true)
        {
            if (this.Map[index].Key == key)
            {
                // Removing an element using open-addressing actually causes a bug,
                // because we may create a hole in the list, and our get() may
                // stop searching early when it reaches this hole.
                this.Map[index] = null;
                this.Size -= 1;
                return;
            }
            index += 1;
            index %= this.Capacity;
        }
    }

    public void ReHash()
    {
        this.Capacity = 2 * this.Capacity;
        Pair[] newMap = new Pair[this.Capacity];

        Pair[] oldMap = this.Map;
        this.Map = newMap;
        this.Size = 0;
        foreach (Pair p in oldMap)
        {
            if (p != null)
            {
                this.Put(p.Key, p.Val);
            }
        }
    }

    public void Print()
    {
        foreach (Pair p in this.Map)
        {
            if (p != null)
            {
                Console.WriteLine(p.Key + " " + p.Val);
            }
        }
    }
}
